Flaky test: TestUpgradingPDAndTSOClusters

## Flaky Test
### Which jobs are failing
```text
panic: test timed out after 5m0s
	running tests:
		TestUpgradingPDAndTSOClusters (5m0s)

goroutine 26264 [running]:
testing.(*M).startAlarm.func1()
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/testing/testing.go:2682 +0x605
created by time.goFunc
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/time/sleep.go:215 +0x45

goroutine 1 [chan receive, 5 minutes]:
testing.(*T).Run(0xc000602a80, {0x67d5446, 0x1d}, 0x71216e0)
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/testing/testing.go:2005 +0x9fe
testing.runTests.func1(0xc000602a80)
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/testing/testing.go:2477 +0x86
testing.tRunner(0xc000602a80, 0xc0012179d8)
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/testing/testing.go:1934 +0x21d
testing.runTests(0xc00054b1e8, {0x9916a40, 0x8, 0x8}, {0x170cd93?, 0xc001217a58?, 0x9a28fe0?})
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/testing/testing.go:2475 +0x96d
testing.(*M).Run(0xc001308000)
	/home/prow/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.7.linux-amd64/src/testing/testing.go:2337 +0xed5
go.uber.org/goleak.VerifyTestMain({0x7821620, 0xc001308000}, {0x9a2af80, 0x9, 0x9})
	/home/prow/go/pkg/mod/go.uber.org/goleak@v1.3.0/testmain.go:53 +0x65
github.com/tikv/pd/tests/integrations/tso.TestMain(...)
	/home/prow/go/src/github.com/tikv/pd/tests/integrations/tso/client_test.go:55
```
### CI link
- https://prow.tidb.net/view/gs/prow-tidb-logs/pr-logs/pull/tikv_pd/10597/pull-unit-test-next-gen-3/2064319827626954752
- https://prow.tidb.net/view/gs/prow-tidb-logs/pr-logs/pull/tikv_pd/10600/pull-unit-test-next-gen-3/2064316587829104640
- https://prow.tidb.net/view/gs/prow-tidb-logs/pr-logs/pull/tikv_pd/10846/pull-unit-test-next-gen-3/2064220361863467008
### Reason for failure (if possible)

### Anything else
- package: `github.com/tikv/pd/tests/integrations/tso`
- test: `TestUpgradingPDAndTSOClusters`
- failure: `panic: test timed out after 5m0s`
- PRs: 10597, 10600, 10846

--- a/tests/integrations/tso/client_test.go
+++ b/tests/integrations/tso/client_test.go
@@ -684,6 +684,9 @@ func TestUpgradingPDAndTSOClusters(t *testing.T) {
 
 	// Create a PD client in microservice env to let the PD leader to forward requests to the TSO cluster.
 	re.NoError(failpoint.Enable("github.com/tikv/pd/client/servicediscovery/usePDServiceMode", "return(true)"))
+	defer func() {
+		re.NoError(failpoint.Disable("github.com/tikv/pd/client/servicediscovery/usePDServiceMode"))
+	}()
 	pdClient, err := pd.NewClientWithContext(ctx,
 		caller.TestComponent,
 		[]string{backendEndpoints}, pd.SecurityOption{}, opt.WithMaxErrorRetry(1))
@@ -693,6 +696,11 @@ func TestUpgradingPDAndTSOClusters(t *testing.T) {
 	// Create a TSO cluster which has 2 servers
 	tsoCluster, err := tests.NewTestTSOCluster(ctx, 2, backendEndpoints)
 	re.NoError(err)
+	defer func() {
+		if tsoCluster != nil {
+			tsoCluster.Destroy()
+		}
+	}()
 	tsoCluster.WaitForDefaultPrimaryServing(re)
 	// The TSO service should be eventually healthy
 	utils.WaitForTSOServiceAvailable(ctx, re, pdClient)
@@ -706,11 +714,9 @@ func TestUpgradingPDAndTSOClusters(t *testing.T) {
 	// Restart the TSO cluster
 	tsoCluster, err = tests.RestartTestTSOCluster(ctx, tsoCluster)
 	re.NoError(err)
-	defer tsoCluster.Destroy()
+	tsoCluster.WaitForDefaultPrimaryServing(re)
 	// The TSO service should be eventually healthy
 	utils.WaitForTSOServiceAvailable(ctx, re, pdClient)
-
-	re.NoError(failpoint.Disable("github.com/tikv/pd/client/servicediscovery/usePDServiceMode"))
 }
 
 func checkTSO(

